### PR TITLE
[Hardware][XPU] AWQ/GPTQ support for xpu backend

### PR DESCRIPTION
**Pull request by yma11**
Add AWQ/GPTQ support for xpu backend.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 1]
```diff
@@ -70,32 +73,118 @@ def get_config_filenames() -> List[str]:
...
+            return cls(method, weight_bits, group_size, False, False)
+        if method == "gptq":
+            weight_bits = cls.get_from_keys(config, ["bits"])
+            group_size = cls.get_from_keys(config, ["group_size"])
+            lm_head_quantized = cls.get_from_keys_or(config, ["lm_head"],
+                                                     default=False)
+            try:
+                desc_act = cls.get_from_keys(config, ["desc_act"])
+            except Exception:
+                desc_act = False
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:89]: Can you just use `cls.get_from_keys_or`?

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 2]
```diff
@@ -70,32 +73,118 @@ def get_config_filenames() -> List[str]:
...
             return cls.get_name()
 
         return None
 
     def get_quant_method(self, layer: torch.nn.Module,
                          prefix: str) -> Optional["LinearMethodBase"]:
         if isinstance(layer, LinearBase):
-            return self.quant_method(self)
+            if self.method == "awq":
+                return IPEXAWQLinearMethod(self)
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:110]: Can you respect awq's modules_to_not_convert?
https://github.com/vllm-project/vllm/blob/de0e61a3239abff67c789138187a98465b806f76/vllm/model_executor/layers/quantization/awq_marlin.py#L123-L124

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 3]
```diff
@@ -70,32 +73,118 @@ def get_config_filenames() -> List[str]:
...
+            if self.method == "gptq":
+                return IPEXGPTQLinearMethod(self)
         return None
 
+    def get_scaled_act_names(self) -> List[str]:
+        if self.method == "awq":
+            return ["gelu", "gelu_fast", "gelu_new", "gelu_pytorch_tanh"]
+        else:
+            return []
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:119]: This function has been removed on main

**Review by mgoin — COMMENTED** [review]
Make sure to run `./format.sh`

File: vllm/model_executor/layers/quantization/gptq.py [hunk 4]
```diff
@@ -210,7 +210,6 @@ def create_weights(
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         # for torch.compile
-        layer.qweight = Parameter(layer.qweight.data, requires_grad=False)
```
[the remark this answers is not in the document]
Comment (yma11) [reply vllm/model_executor/layers/quantization/gptq.py:213]: This is a duplicated [line](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/gptq.py#L213-L215).

**Comment by zhouyuan** [comment]
Please also help to update the doc:
https://github.com/vllm-project/vllm/blob/main/docs/source/quantization/supported_hardware.rst

**Comment by yma11** [comment]
> Please also help to update the doc: https://github.com/vllm-project/vllm/blob/main/docs/source/quantization/supported_hardware.rst

Done.

**Comment by yma11** [comment]
@mgoin Thanks for comments. I updated accordingly. Can you help review again? Thanks.

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 5]
```diff
@@ -70,32 +74,113 @@ def get_config_filenames() -> List[str]:
...
+            if ipex.__version__ < "2.5.0":
+                raise ImportError("intel_extension_for_pytorch version is "
+                                  "wrong. Please install "
+                                  "intel_extension_for_pytorch>=2.5.0.")
+        except ImportError as err:
+            raise ImportError(
+                "Please install "
+                "intel_extension_for_pytorch>=2.5.0 via "
+                "`pip install intel_extension_for_pytorch>=2.5.0`"
+                " to use IPEX-AWQ linear method.") from err
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:140]: Since you are using "2.5.0" for all minimum version checks in this file, could you store it as a variable and reference it?
Something like `MIN_IPEX_VERSION = "2.5.0"` at the top - then you can refactor here and below:
```suggestion
            if ipex.__version__ < MIN_IPEX_VERSION:
                raise ImportError("intel_extension_for_pytorch version is "
                                  "wrong. Please install "
                                  f"intel_extension_for_pytorch>={MIN_IPEX_VERSION}.")
        except ImportError as err:
            raise ImportError(
                "Please install "
                f"intel_extension_for_pytorch>={MIN_IPEX_VERSION} via "
                f"`pip install intel_extension_for_pytorch>={MIN_IPEX_VERSION}`"
                " to use IPEX-AWQ linear method.") from err
```

File: vllm/model_executor/model_loader/loader.py [hunk 6]
```diff
@@ -386,7 +386,7 @@ def load_model(self, vllm_config: VllmConfig) -> nn.Module:
 
             for _, module in model.named_modules():
                 quant_method = getattr(module, "quant_method", None)
-                if quant_method is not None:
+                if isinstance(quant_method, QuantizeMethodBase):
```
Comment (mgoin) [inline vllm/model_executor/model_loader/loader.py:389]: Why was this change needed rather than just a None check?

**Review by mgoin — COMMENTED** [review]
LGTM overall, just one more suggestion

File: vllm/model_executor/model_loader/loader.py (hunk 6 above)
Comment (yma11) [reply vllm/model_executor/model_loader/loader.py:324]: When iterating the `named_modules` on intel platforms, a customed layer in intel IPEX actually doesn't have a valid `quant_method` even it's not `None`. So I tighten up this check here to skip processing it which is not necessary.

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 5 above)
Comment (yma11) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: Thanks for suggestion. Updated.

**Review by mgoin — APPROVED** [review]
Thank you for all the changes, this looks great!
